File: android/app/src/main/java/info/bagen/rust/plaoc/microService/sys/http/net/Http1Server.kt
package info.bagen.rust.plaoc.microService.sys.http.net

import info.bagen.rust.plaoc.microService.helper.PromiseOut
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.http4k.core.HttpHandler
import org.http4k.server.Http4kServer
import org.http4k.server.Netty
import org.http4k.server.asServer


class Http1Server {
    companion object {
        const val PREFIX = "http://";
        const val PROTOCOL = "http:";
        const val PORT = 80;
    }

    var bindingPort = -1

    private var server: Http4kServer? = null
    suspend fun createServer(handler: HttpHandler) {
        if (server != null) {
            throw Exception("server alter created")
        }

        val portPo = PromiseOut<Int>()
        CoroutineScope(Dispatchers.IO).launch {
            server = handler.asServer(Netty(0/* 使用随机端口*/)).start().also { server ->
                bindingPort = server.port()
                portPo.resolve(bindingPort)
            }
        }
        portPo.waitPromise()
    }

    val origin get() = "${PREFIX}localhost:${bindingPort}"

    fun closeServer() {
        server?.also {
            it.close()
            server = null
        } ?: throw Exception("server not created")
    }
}

File: next/dweb-browser/MicroService/src/Sys/Http/Net/Http1Server.cs

namespace DwebBrowser.MicroService.Sys.Http.Net;

public class Http1Server
{
    public readonly static string PREFIX = "http://";
    public readonly static string PROTOCOL = "http:";
    public readonly static int PORT = 80;

    private int _bindingPort = -1;

    private IServerInfo<HttpListener>? _info;

    public IServerInfo<HttpListener>? Info
    {
        get => _info;
    }

    public string Authority
    {
        get => string.Format("localhost:{0}", _bindingPort);
    }

    public string Origin
    {
        get => string.Format("{0}{1}", PREFIX, Authority);
    }

    public IServerInfo<HttpListener> CreateServer(HttpHandler handler, WebSocketHandler webSocketHandler)
    {
        int local_port = _bindingPort = PortHelper.FindPort(new int[] { 22605 });

        return NetServer.HttpCreateServer(new ListenOptions(local_port), handler, webSocketHandler);
    }

    public void CloseServer()
    {
        _info?.Server.Stop();
        _info = null;
    }
}

